### PR TITLE
[#96] Model batch invocation

### LINKED ISSUE #96: Model invoke batch processing

### PR DESCRIPTION
**Pull request by kirillmakhonin**
This closes #96 

Model batch invocation POST endpoint has been added. It requests urlencoded data lines for each invocation in request's body

### CODE REVIEW

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
test this please

ср, 20 июн. 2018 г., 18:46 legion-bot <notifications@github.com>:

**Comment by kirillmakhonin** [comment]
test this please

File: legion/legion/serving/pyserve.py [hunk 1]
```diff
@@ -86,6 +97,26 @@ def model_invoke(model_id):
     return legion.http.prepare_response(output)
 
 
+@blueprint.route(SERVE_BATCH.format(model_id='<model_id>'), methods=['POST'])
+def model_batch(model_id):
+    """
+    Call model for calculation in batch model
```
Comment (aliaksandr-d) [inline legion/legion/serving/pyserve.py]: in batch mode

File: legion/legion/serving/pyserve.py [hunk 2]
```diff
@@ -86,6 +97,26 @@ def model_invoke(model_id):
...
+
+    :param model_id: model name
```
Comment (aliaksandr-d) [inline legion/legion/serving/pyserve.py]: You pass a Model Name to validate_model_id function. It confuses.

File: legion/tests/test_pyserve.py [hunk 3]
```diff
@@ -142,6 +143,17 @@ def test_model_invoke_summation_with_df_and_get_post_method(self):
             self.assertIsInstance(result, dict, 'Result not a dict')
             self.assertDictEqual(result, {'x': a + b})
 
+    def test_model_invoke_summation_with_batch_mode(self):
```
Comment (aliaksandr-d) [inline legion/tests/test_pyserve.py]: may be *in_batch_mode*?

File: legion/tests/test_pyserve.py [hunk 4]
```diff
@@ -142,6 +143,17 @@ def test_model_invoke_summation_with_df_and_get_post_method(self):
...
+        with ModelServeTestBuild(self.MODEL_ID, self.MODEL_VERSION,
+                                 create_simple_summation_model_by_df) as model:
+            parameters = [{'a': randint(1, 20), 'b': randint(2, 50)} for _ in range(10)]
+            expected_answer = [{'x': pair['a'] + pair['b']} for pair in parameters]
+
+            response = model.model_client.batch(parameters)
+
+            self.assertIsInstance(response, list, 'Result not a list')
```
Comment (aliaksandr-d) [inline legion/tests/test_pyserve.py]: _Response is not a list_

File: legion/tests/test_pyserve.py [hunk 5]
```diff
@@ -200,10 +212,33 @@ def test_model_invoke_summation_with_untyped_columns(self):
             self.assertIsInstance(result, dict, 'Result not a dict')
```
Comment (aliaksandr-d) [inline legion/tests/test_pyserve.py]: Result is no a dict

File: legion/legion/serving/pyserve.py (hunk 2 above)
Comment (kirillmakhonin) [reply legion/legion/serving/pyserve.py]: why?

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by aliaksandr-d** [comment]
Looks good for me

**Comment by kirillmakhonin** [comment]
Documentation should be added

**Comment by aliaksandr-d** [comment]
@akharlamov Please, review. All issues are fixed. Thank you!

File: legion/legion/http.py [hunk 6]
```diff
@@ -19,15 +19,41 @@
...
 
 import flask
+from requests.compat import urlencode
+from requests.utils import to_key_val_list
+from urllib.parse import parse_qs
 
 LOGGER = logging.getLogger(__name__)
 
 
+def encode_http_params(data):
```
Comment (akharlamov) [inline legion/legion/http.py]: Should not this be in the client module?

Comment (kirillmakhonin) [reply legion/legion/http.py]: Yeah, but model client is a part of legion package, so all functionalities related to HTTP protocol placed here.

**Pull request merged**